ENH: Deprecate writeable broadcast_array

mattip opened the pull request:
~Relates to #2705, will fix it when deprecation warning is removed and readonly flag set. Replaces #11667.~

As suggested in #11667 and in the code, the view returned from `broadcast_arrays` should be made `readonly`. ~set the `NPY_ARRAY_WARN_ON_WRITE` flag if the data is repeated across the array to fit the new shape.~ Currently we leave the `readonly` flag `False` which can lead to errors.

~An alternative would be to immediately set the `readonly` flag False, with no deprecation cycle. Advantages to the alternative path:~

~- we are disrupting workflows anyway, we could do it only once instead of twice (once to filter the `DeprecationWarning`, once to remove the filter and handle the `readonly` change)~
~- The current behavior is clearly wrong and will cause bugs if the returned arrays are written to.~

I will solicit responses from the mailing list

Edit: modified to set `NPY_ARRAY_WARN_ON_WRITE`

Edit: modified to set a DeprecationWarning

mattip committed:
ENH: Deprecate writeable broadcast_array

seberg commented:
Was going to answer on the list, but another (or two) emails of me are not being accepted by the mailing list right now:

We have/used to have `NPY_ARRAY_WARN_ON_WRITE` flag for some deprecations, changing the warning text to include broadcast arrays, may be a way to have at least some deprecation cycle, that does not require any changed. I am not quite sure it fits here, but if it does and we are not in a rush, it seems like a reasonable thing to do that first.

seberg commented:
The diagonal stuff is a bit cursed IIRC. @njsmith can probably give a quick hint on were it got stuck, I can't say I fully remember right now. I seem to recall that there was a lot of pushback or simply problems around the whole thing, I think it was one of the examples for "deprecations that did not turn out to go well" in the deprecations NEP or roadmap thingy.

ahaldane commented:
I can comment on diag and `NPY_ARRAY_WARN_ON_WRITE`, as it is related to the multifield view stuff that finally got merged in 1.14 after many years in #12447. In that PR I guess I got rid of the last code that set the flag, apparently `np.diagonal` did not do so and instead used a readonly flag.

The original plan was to make both diag and multifield-indexes return views instead of (packed) copies. Part of the problem is that this breaks views of the result, because the views in both cases have padding bytes, but the copy does not. 

See #7661 #5407 #5409

Now that the multifield stuff is in I guess we could either remove or repurpose `NPY_ARRAY_WARN_ON_WRITE`.

mattip commented:
The [mailing list](http://numpy-discussion.10968.n7.nabble.com/Warn-or-immidiately-change-readonly-flag-on-broadcast-arrays-return-value-td46560.html) generated a few comments, with a request to not change the API as it understood that this may create overlapping memory. 

The [`broadcast_to`](https://github.com/numpy/numpy/blob/v1.16.2/numpy/lib/stride_tricks.py#L144) api return readonly views.

Perhaps repurposing [`array_might_be_written`](https://github.com/numpy/numpy/blob/v1.16.2/numpy/core/src/multiarray/arrayobject.c#L651) with a more specific warning would be the best path, such a warning would be easy to filter for experts who don't want to see it.

mhvk commented:
Belatedly: I'd be quite happy to just go directly to changing the default to `readonly=True`, but if a warning on writing to a broadcast array with zero strides is not too much hassle, that's obviously OK too. Just as long as we don't warn everybody who uses the function in effectively readonly fashion.

mattip commented:
I have repurposed `NPY_ARRAY_WARN_ON_WRITE` for general use. In order to actually use it in `broadcast_array`, I needed to expose it to python like any other flag. Does this need to hit the mailing list as an API change?

ahaldane commented:
I guess the only question in my mind is whether it is right to make this flag a part of the python API, because that means we may need to keep supporting it forever. Eg, if someone writes a library which turns off the flag, that code needs to work even after we no longer use the flag anywhere.

But, I think that is acceptable, and the flag seems useful to have around in case similar cases turn up in the future. I, for one, am fine merging without discussion on the list - this isn't going to break anyone's code.

seberg commented:
Oh, have to think about fully exposing it/this a bit more I guess. I somewhat thought we would re-purpose it for deprecation only (switching over to giving an error on such writes, unless someone opts in with `writeable=True`. Which would be OK with a private function).
On the other hand, if we start giving lots of errors due to not writeable arrays, it might be good have a clearer error message as to why some array is suddenly not writeable.

seberg commented:
OK, after discussing with Stephan Hoyer, I think we should morph this into:

* Simply deprecate `np.broadcast_arrays` returning a writeable view
* Give a warning saying that you will have to do `arr.flags.writeable = True` if you want to keep the old behaviour. We could point out that `broadcast_arrays` to add a singleton dimension is not too helpful.
* If too many people complain, we expose a `readonly=True` kwarg to `np.broadcast_arrays`.

I do remember that there was some push-back on the mailing list, but IIRC it was mostly one person and it was a bit unclear whether there was a real use-case around it.

I do think we should not expose new public API for this, instead have a private function to set the flag (since we need it).

mattip commented:
Needs #13463 to fix a bug with `_IsWritable` before this can go in

seberg committed:
ENH: Make writeable flag enabling more reliable for non-array bases

When the base is not an array (or generally when the flag of the base
array was toggled), it is OK to allow setting the writeable flag to
True, as long as any ancestor (especially the last one) is writeable.

seberg commented:
Rebased. I forgot to add that fix to gh-13463, inserted it as a new commit here now, but maybe it would be better as its own PR. I hope you do not mind the forced pushing.

mhvk commented on doc/release/1.17.0-notes.rst:
```diff
@@ -42,6 +42,18 @@ In the future it will not be possible to switch the writeable flag to ``True``
+
+Commonly `numpy.broadcast_array` returns a writeable array with internal
+overlap, making it unsafe to write to. A future version will set the
+``writeable`` flag ``False``, and require users to manually set it ``True`` if
```
set *to* `False` and set it *to* `True`.

mattip replied:
adopted and re-flowed the paragraph

mhvk commented on doc/release/1.17.0-notes.rst:
```diff
@@ -42,6 +42,18 @@ In the future it will not be possible to switch the writeable flag to ``True``
+Commonly `numpy.broadcast_array` returns a writeable array with internal
+overlap, making it unsafe to write to. A future version will set the
+``writeable`` flag ``False``, and require users to manually set it ``True`` if
+they are sure that is what they want to do. Writing to it will now emit a
```
Maybe start sentence with "For the 1.17 version, writing to it will emit a deprecation warning... *to* `True`.

mattip replied:
adopted

mhvk commented on doc/release/1.17.0-notes.rst:
```diff
@@ -42,6 +42,18 @@ In the future it will not be possible to switch the writeable flag to ``True``
+``writeable`` flag ``False``, and require users to manually set it ``True`` if
+they are sure that is what they want to do. Writing to it will now emit a
+deprecation warning with instructions to set the ``wrietable`` flag ``True``.
+This creates a confusing situation where even though the ``writeable`` flag is
```
I think this sentence only confuses the reader more! Maybe instead  "Note that if one were to inspect the flag before setting it, one would, perhaps confusingly, find it would already be `True`. Explicitly setting it, though, as one will need to do in future versions, clears an internal flag that is used to produce the deprecation warning."

mattip replied:
adopted

mhvk commented on numpy/core/src/multiarray/flagsobject.c:
```diff
@@ -207,15 +208,36 @@ arrayflags_dealloc(PyArrayFlagsObject *self)
         return item; \
     }
 
+static char * msg = "future versions will not create a writeable "
```
Typically we do not seem to have a space after the `*`

mhvk commented on numpy/core/src/multiarray/flagsobject.c:
```diff
@@ -315,7 +337,8 @@ arrayflags_updateifcopy_set(PyArrayFlagsObject *self, PyObject *obj)
     res = PyObject_CallMethod(self->arr, "setflags", "OOO", Py_None, Py_None,
-                              (PyObject_IsTrue(obj) ? Py_True : Py_False));
+                              (PyObject_IsTrue(obj) ? Py_True : Py_False)
+                              );
```
why this change?

mattip replied:
reverted

mhvk commented on numpy/core/src/multiarray/flagsobject.c:
```diff
@@ -339,8 +362,9 @@ arrayflags_writebackifcopy_set(PyArrayFlagsObject *self, PyObject *obj)
-                              (PyObject_IsTrue(obj) ? Py_True : Py_False));
+    res = PyObject_CallMethod(self->arr, "setflags", "OOO0", Py_None, Py_None,
+                              (PyObject_IsTrue(obj) ? Py_True : Py_False),
+                              Py_None);
```
Is this addition still necessary? Below you comment that you're not mangling `setflags` (I'm probably just confused...)

mattip replied:
reverted cruft left over from intermediate versions

mhvk reviewed (commented):
This looks good but I think the change log entry could be a bit clearer; hopefully my suggestions make sense. The rest is mostly nitpicks, though also one query.

charris commented:
> The diagonal stuff is a bit cursed IIRC

If I recall, we converted it to a view and made it readonly with intent to change that later. At that point I'm forgetful, although ISTR that there was a reason I didn't precede to finish that program.

mattip commented on doc/release/1.17.0-notes.rst:
```diff
@@ -129,6 +129,24 @@ New mode "empty" for ``np.pad``
+When `numpy.broadcast_array` can return a writeable array with internal overlap,
+it will set the ``NPY_ARRAY_WARN_ON_WRITE`` flag `True` to reduce
+unintentionally writing to more than one location by mistake in the returned
+array.
```
reworded, please review the newer version

mhvk commented on doc/release/1.17.0-notes.rst:
```diff
@@ -42,18 +42,21 @@ In the future it will not be possible to switch the writeable flag to ``True``
-behaviour. An additional `FutureWarning` will be emitted when accessing the
-``writeable`` flag state to clarify the contradiction.
+``writeable`` flag to ``False``, and require users to manually set it to
+``True`` if they are sure that is what they want to do. From this version,
```
From this version -> In this version or perhaps "Until then, " depending on what the goal was.

mattip replied:
Now, writing to it ...

mhvk commented on doc/release/1.17.0-notes.rst:
```diff
@@ -43,6 +43,21 @@ from python.
+setting it, one would, perhaps confusingly, find it would already be ``True``.
+Explicitly setting it, though, as one will need to do in future versions,
+clears an internal flag that is used to produce the deprecation warning. An
+additional `FutureWarning` will be emitted when accessing the ``writeable``
```
Upon looking at the PR again, I now see where this second future warning is emitted. Maybe delete the "perhaps confusingly" insert I suggested earlier and start this sentence with: "To help alleviate confusion, an additional ..."

mhvk reviewed (commented):
This now looks great. A few more wording comments for the changelog entry - hopefully me working out my confusing will end up helping others not be confused...

mhvk reviewed (approved):
OK, all great now!

seberg commented:
Thanks Marten, will plan to make a quick pass and then merge it today probably. Should probably add my first commit to the realease notes as well, but will do that in a followup PR maybe, seems easier at this point.

mhvk commented:
@seberg - would seem fine to add something to the release notes here, if you feel it needs it. Actually, should it be an ENH or BUG prefix?

seberg commented on numpy/core/src/multiarray/flagsobject.c:
```diff
@@ -623,7 +677,7 @@ arrayflags_setitem(PyArrayFlagsObject *self, PyObject *ind, PyObject *item)
         return arrayflags_updateifcopy_set(self, item);
     }
-    else if (((n==14) && (strncmp(key, "WRITEBACKIFCOPY", n) == 0)) ||
+    else if (((n==15) && (strncmp(key, "WRITEBACKIFCOPY", n) == 0)) ||
```
Sneaky bug fix ;). (lets see, maybe I will add tests for this).

mhvk commented on doc/release/1.17.0-notes.rst:
```diff
@@ -129,6 +129,15 @@ the stream changes for any given seed is extremely small. If a 0 is encountered
 underlying generator, then the incorrect value produced (either ``np.inf``
 or ``np.nan``) is now dropped.
 
+``arr.writeable`` can be set to true slightly more liberal
```
liberal -> liberally

seberg replied:
had missed this, changed it to read "``arr.writeable`` can be switched to true slightly more often", thought that might be simpler english.

mhvk commented on doc/release/1.17.0-notes.rst:
```diff
@@ -129,6 +129,15 @@ the stream changes for any given seed is extremely small. If a 0 is encountered
+``arr.writeable`` can be set to true slightly more liberal
+----------------------------------------------------------
+
+In rare cases, it can now be possible to switch an array from not writeable
```
I'd turn this around and state the problem first, like
"In rare cases, it was not possible to switch an array from not writeable 
to writeable, even if the base was writeable, because the search for the
base array had a bug. This is now fixed."

Or is this not true?

seberg replied:
Half true, I made it now:
```
In rare cases, it was not possible to switch an array from not writeable
to writeable, although a base array is writeable. This can happen if an
intermediate ``arr.base`` object is writeable. Previously, only the deepest
base object was considered for this decision. However, in rare cases this
object does not have the necessary information. In that case switching to
writeable was never allowed. This has now been fixed.
```
(will push in a minute)

mhvk replied:
I'm still slightly confused, but I think would get the point that what was a problem is now fixed!

seberg replied:
It is a bit confusing. Basically, we do `arr.base.base.base` (until reaching the end, or a non-array). If the last one is not an array, "readonly" logic only works for buffers, not for `__array_interface__` or so. The change, is to check `writeable` on every base and stop early. Which fixes the case when the last base object is an object where we cannot find out whether or not it is writeable.

mhvk commented on numpy/core/src/multiarray/flagsobject.c:
```diff
@@ -702,7 +707,8 @@ static PyObject *
 {
     int fl = self->flags;
-    char * warn_on_write="";
+    char * warn_on_write = "";
```
As you're touching this anyway, no space after `*`?

seberg replied:
Haha, oops, fixed :)

mhvk reviewed (commented):
Nitpicks that you can ignore if you wish...

mattip commented:
@seberg ping

seberg commented:
@mattip, I will have another look over (and rebase). I guess my last fixups will have satified Marten, so I/we can merge (may just do later). If someone wants more tests here, please just ping me and I will add them in a followup in that case.

mhvk commented:
Yes, all looks good to me.

eric-wieser commented on numpy/core/src/multiarray/common.c:
```diff
@@ -632,7 +632,17 @@ _IsWriteable(PyArrayObject *ap)
+
         if (base == NULL || PyArray_CHKFLAGS(ap, NPY_ARRAY_OWNDATA)) {
+            /* there is no further base to test the writeable flag for */
             return (npy_bool) PyArray_ISWRITEABLE(ap);
```
I think it would be clearer if this were `return False`, since at this point `ap` can never be writeable

mattip replied:
adopting

eric-wieser commented on numpy/core/src/multiarray/flagsobject.c:
```diff
@@ -207,15 +208,36 @@ arrayflags_dealloc(PyArrayFlagsObject *self)
+        }\
+        item = ((self->flags & (UPPER)) == (UPPER)) ? Py_True : Py_False; \
+        Py_INCREF(item); \
+        return item; \
```
nit: could combine these three lines and the declaration to `return PyBool_FromLong((self->flags & (UPPER)) == (UPPER))` (same on line 206). But I suppose this is consistent with what was there before

mattip replied:
adopting

eric-wieser commented on numpy/core/src/multiarray/flagsobject.c:
```diff
@@ -398,6 +420,35 @@ arrayflags_writeable_set(PyArrayFlagsObject *self, PyObject *obj)
+                "Cannot delete flags _warn_on_write attribute");
+        return -1;
+    }
+    if (PyObject_IsTrue(obj)) {
```
PyObject_IsTrue can return `-1`, you need to check for that

mattip replied:
fixing

eric-wieser commented on numpy/core/src/multiarray/flagsobject.c:
```diff
@@ -398,6 +420,35 @@ arrayflags_writeable_set(PyArrayFlagsObject *self, PyObject *obj)
+    }
+    else {
+        PyErr_SetString(PyExc_ValueError,
+                        "cannot set warn_on_write flag to False, set "
```
Shouldn't this name include the underscore? So `cannot clear _warn_on_write flag ...`?

mattip replied:
adopting

eric-wieser commented on numpy/core/src/multiarray/flagsobject.c:
```diff
@@ -648,19 +707,25 @@ static PyObject *
 arrayflags_print(PyArrayFlagsObject *self)
 {
     int fl = self->flags;
+    char *warn_on_write = "";
```
`const char *`?

mattip replied:
adopting

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -222,8 +222,14 @@ def broadcast_arrays(*args, **kwargs):
-        broadcasted array may refer to a single memory location.  If you
-        need to write to the arrays, make copies first.
+        broadcasted array may refer to a single memory location. If you need
+        to write to the arrays, make copies first or set the writable flag
```
Perhaps double backticks around ``writeable``

mattip replied:
adopting

eric-wieser commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -415,12 +415,27 @@ def test_writeable():
         for result in results:
-            assert_equal(result.flags.writeable, True)
+            # This will change to False in a future version
+            if any([s == 0 for s in result.strides]):
```
Simpler as `if 0 in result.strides`?

eric-wieser replied:
Also, should there be an `else` on this test? What behavior do we expect if no dimensions were broadcast?

I'd be inclined to keep the result warn-on-writeable for consistency.

mattip replied:
reworked the test. Since `ret = broadcast_arrays([a])` will return `ret[0] is a`, we cannot always set the `warn_on_writeable` flag, that would mean changing the original array.

eric-wieser replied:
Does `boradcast_arrays(a, a)` return `a` twice by identity?

mattip replied:
Yes
```
>>> res = np.broadcast_arrays(a, a)
>>> res[0] is a
True
>>> res[1] is a
True
```

eric-wieser commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -415,12 +415,27 @@ def test_writeable():
+        for result in results:
+            # resets the warn_on_write DeprecationWarning
+            result.flags.writeable = True
+            result[:] = 0
```
Can you try accessing `result.flags.writeable` here too, and prove it raises no warning?

mattip replied:
ok

eric-wieser commented on doc/release/1.17.0-notes.rst:
```diff
@@ -43,6 +43,21 @@ from python.
+Writing to the result of `numpy.broadcast_array` will warn
+----------------------------------------------------------
+
+Commonly `numpy.broadcast_array` returns a writeable array with internal
```
Plural `broadcast_arrays`

mattip replied:
fixing

eric-wieser commented:
Approach looks sounds, but some of the prose seems wrong, and there's an unhandled error path.

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2563,7 +2565,6 @@ array_setflags(PyArrayObject *self, PyObject *args, PyObject *kwds)
             PyArray_CLEARFLAGS(self, NPY_ARRAY_WRITEABLE);
```
Should this clear warn_on_write too? If the caller sets the result of `broadcast_arrays` to have `result.flags.writeable = False`, it seems to me that the warning is not needed.

eric-wieser commented:
I assume the flag gets propagated?
```python
a, b = np.broadcast_arrays(np.arange(3), np.arange(3)[:,None])
a.writeable  # warns
a2 = a.view()
a2.writeable  # but does this warn?
```

mattip commented:
Yes, the flag propagates. However, if after creating `a2`, the flag on `a` is reset, `a2` still warns. I don't know if changing that is worth the effort for a deprecation on a rarely-used function.

<details>

```
>>> a, b = np.broadcast_arrays(np.arange(3), np.arange(3)[:,None])
>>> a.flags.writeable  # warns
__console__:1: FutureWarning: future versions will not create a writeable array from broadcast_array. Set the writable flag explicitly to avoid this warning.
True
>>> a2 = a.view()
>>> a2.flags.writeable
__console__:1: FutureWarning: future versions will not create a writeable array from broadcast_array. Set the writable flag explicitly to avoid this warning.
True
>>> a.flags.writeable = True
>>> a2.flags.writeable
__console__:1: FutureWarning: future versions will not create a writeable array from broadcast_array. Set the writable flag explicitly to avoid this warning.
True
```

eric-wieser commented:
Agreed that we don't need to fix that. Maybe worth documenting that behavior somehow, either just in the release notes or the warning message. Something about fixing up the flag before taking views such as by slicing.

mattip commented:
@eric-wieser, @seberg: I rebased and squashed this. Any more review?

seberg commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -415,12 +415,29 @@ def test_writeable():
             assert_equal(result.flags.writeable, True)
+            result[:] = 0
+            
+                
```
```suggestion
```

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -191,6 +224,16 @@ def test_writeable_from_c_data(self):
+            a[2] = 10
+            # only warn once
+            assert_(len(w) == 1)
+
```
```suggestion
```

seberg reviewed (approved):
Didn't go through it super carefully, but I remember it being all fine. And tests seem thorough enough, some blank lines are missing in the release notes, but I can also do that later. Just in case someone still wants to comment, not merging it right now.

mattip committed:
Update doc/release/1.17.0-notes.rst

Co-Authored-By: Sebastian Berg <sebastian@sipsolutions.net>

mattip committed:
Update doc/release/1.17.0-notes.rst

Co-Authored-By: Sebastian Berg <sebastian@sipsolutions.net>

mattip committed:
Update numpy/lib/tests/test_stride_tricks.py

Co-Authored-By: Sebastian Berg <sebastian@sipsolutions.net>

mattip committed:
Update numpy/core/tests/test_multiarray.py

Co-Authored-By: Sebastian Berg <sebastian@sipsolutions.net>

mattip committed:
DOC: improve warning (from review)

seberg commented:
Sorry, and thanks for the additional ping. I will put this in.

mhvk commented:
With cython code, this beatifully constructed PR sadly is giving false positives; see #13929
